### ISSUE TITLE
Know when a site is available as a PWA

### ISSUE THREAD

**Issue by pocmo** [mozilla-mobile/fenix#768]
### User Story
As a user, I want to be notified when a website I visit is available as a PWA, so I can install it as a short-cut on my homescreen. 

### Requirements
This is about notifying the user when a site has a web app manifest.

### Acceptance Criteria
- [ ] When I navigate to a site that is available as a PWA, my browser lets me know that I can install the PWA.
- [ ] The way my browser notifies me is not intrusive to my browsing experience.
- [ ] Telemetry Ticket TBD
- [ ] This does NOT include the PWA-app splash screen in the mock below

Abstract: https://share.goabstract.com/3815ff0e-5b89-4a41-a1de-7414f2e57964?mode=design&sha=63bb397f7bc0a0c98f7d225f90c092026dd264e5

**Note to ENG from Vesta:** 

-Most ENG work related to basic implementation will be completed by Tiger in Q4. This should be the only remaining PWA ticket from the original META #204 and it is focused on the UI aspect of the user experience
-Please ignore the splash screen in the mocks for now as it is out of scope for Q1 :)
-Feature brief for reference: https://docs.google.com/document/d/1Dd0OnivaUMe6rvEP5yi6Gt7Hz785ZstDLytaxnXz7Dk/edit

**Comment by reinhart1010** [comment]
Duplicate: https://groups.google.com/forum/m/?utm_medium=email&utm_source=footer#!topic/fenix-nightly/y0s-eiLwA0o

**Comment by apbitner** [comment]
Description updated with Abstract link showing First Use and Toolbar Notification flows

**Comment by madb1lly** [comment]
This is a nice mockup, but what is the reason that the icon cannot be put in the toolbar like in Fennec? Is there less space in Fenix?

Also, "install" may be confusing as these are not really installed in the traditional Android sense - they will not appear in the list of installed apps. This could be mitigated by Fenix having a record of installed PWAs, e.g. "about:pwas" or "about webapps" where they can be managed. Firefox Accounts could also keep a record so that they can easily be reinstalled.

**Comment by apbitner** [comment]
@madb1lly we have a handful of features that are only available depending on the website you're on such as reader mode, open in native app, install PWA, etc. We don't want to put all of those icons in the address bar so we're working on a way to make these contextual features easily noticeable without cluttering up the address bar.

**Comment by liuche** [comment]
@NotWoods my understanding here is that once the notification dot is done, this bug will be done? If so I'm removing it from this board.

**Comment by NotWoods** [comment]
Yes, although I need to integrate it with Fenix after mozilla-mobile/android-components#5122 is merged.

**Comment by madb1lly** [comment]
Hi @apbitner, all those things you list (reader mode, open in native app, install PWA) fit in the address bar in Fennec so I'm struggling to understand the difficulty of doing this in Fenix. Or is just that the Fennec solution isn't considered to meet Fenix's UX guidance?

In any case, I still think "install" to homescreen could be considered misleading without including some way to record installed PWAs within Fenix itself (since Android will keep no backup of this).

**Comment by apbitner** [comment]
@madb1lly we still plan on taking a look at this - https://github.com/mozilla-mobile/fenix/issues/6350

**Comment by NotWoods** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Implement PWA install UI](https://trello.com/c/gB2lNXNF/642-implement-pwa-install-ui)

**Comment by AndiAJ** [comment]
Hi, I've just tested this issue on the latest Nightly Build #13540607from 12/20 using the following devices:
• Google Pixel 3a (Android 9)
• Huawei Mate 20 Lite (Android 8.1.0)
• Samsung Galaxy S7 (Android 7)
• OnePlus A3 (Android 6.0.1)
• LG Nexus 4 (Android 5.1.1)

After while testing and double checking against the **[Abstract](https://app.goabstract.com/share/3815ff0e-5b89-4a41-a1de-7414f2e57964?mode=design&sha=9cb28aba843a30c0d5dd31ac29088c2664eb1898)** I've notied a couple of things:

1. "Install" String instead of "Install on Home screen" (as per the spec)
![rsz_11rsz_screenshot_20191220-135848](https://user-images.githubusercontent.com/51314259/71257329-1d9c1680-233c-11ea-9000-7b1123eb12d1.png)

2. The user has to close the tab and re-open it again to get the 3dot button toolbar notification.
    If you resume the page from the Open Tabs section it will not be displayed again.

3. How many times can a user add the same PWA to the home screen?
   I've noticed that on the Pixel, OnePlus & LG Nexus I can add the same PWA as many times as I want.
  On the Huawei, and Samsung I can only add it to the homescreen once.
► Video
![20191220_150942](https://user-images.githubusercontent.com/51314259/71258197-3efe0200-233e-11ea-9805-ac89be746e85.gif)

4. On the OnePlus and LG Nexus the Android Dialog box isn't displayed.
    On the other hand when adding a Website Shortcut the dialog is properly displayed
► Video
![20191220_153051](https://user-images.githubusercontent.com/51314259/71258411-c51a4880-233e-11ea-9087-cce4091b7cfd.gif)


► Full check on Google Pixel 3a
![20191220_150420](https://user-images.githubusercontent.com/51314259/71258766-a4062780-233f-11ea-8390-b5de5b275e59.gif)


@NotWoods & @apbitner  - Please review, and advise. ☺️ 
I'll remove the QA needed label until further notice.

**Comment by apbitner** [comment]
1 - Good catch. 'Install' is correct, the design just hadn't been updated, but I'll do so to avoid any confusion.

For 2 - 4, these sound like they could be bugs so I'll let @NotWoods weigh in.

**Comment by NotWoods** [comment]
2 is a bug, 3 we can't control. 

With 4, we currently display the native shortcut dialog. On older devices, there is no dialog to display. We could add some variation of the bookmark shortcuts dialog for all devices, but then we would show two dialogs on newer devices.

**Comment by TitanNano** [comment]
@NotWoods just wanted to mention here that this native shortcut dialog is provided by the launcher and not the system UI, so not every launcher actually implements such a dialog.

**Comment by NotWoods** [comment]
I wasn't aware of that, thanks for the info 😄 I'll try to test on a few different launchers.

**Comment by vesta0** [comment]
Hey @NotWoods  is this ready for QA yet?

**Comment by NotWoods** [comment]
No, the dialog hasn't been merged yet

**Comment by AndiAJ** [comment]
Hi, I've re-checked this matter on the latest **`Nightly Build #20940609 from 4/3`** using the following devices:
• Google Pixel 3a (Android 10)
• Huawei Mate 20 Lite (Android 9)
• OnePlus A3 (Android 6.0.1)

The PWA dialog is displayed and seems to be working **only on the first PWA page**.
**Afterwards, no dialogs are displayed if you navigate to other PWA pages for the first time**.
 
The most important aspect is that, **whether you do or don't add a PWA , the "Install" button from the ⋮ Main menu seems to disappear making Fenix incapable of installing any PWA's**.

► Used "Add to home screen" from the dialog and afterwards re-checked the ⋮ Main menus and dialogs 
![20200403-141649](https://user-images.githubusercontent.com/51314259/78360407-7035d180-75bf-11ea-863c-4fb8e51b7b1f.gif)


► Used "Install button"  and afterwards re-checked the ⋮ Main menus and dialogs 
![20200403-142121](https://user-images.githubusercontent.com/51314259/78360147-01587880-75bf-11ea-96e5-d585d9ade119.gif)


► Didn't install any PWA just checked the dialogs and ⋮ Main menu "Install" buttons
![20200403-140003](https://user-images.githubusercontent.com/51314259/78359389-affbb980-75bd-11ea-8308-42d2bb745f91.gif)

@NotWoods & @ekager - Please review.
I'll remove the QA needed.

**Comment by ekager** [comment]
The above issue has been fixed by the fix for #9692 ! I think we can close this. Thanks!
